### DISCUSSION TITLE
AMQP 1.0 [Shovel] Source

### DISCUSSION THREAD

**Discussion by DarkIkki** [rabbitmq/rabbitmq-server#7868] [Other]
### Is your feature request related to a problem? Please describe.

Hi, 
In production RabbitMQCluster v3.10.5
(but the problem is the same in RabbitMQ Server 3.11.13)

I want receive in RabbitMQ (AMQP 0.9.1 target) messages coming from IBM MQ v9.2 with "AMQP 1.0" connection channel (AMQP 1.0 source)
The AMQP 1.0 message subject is not copied in RabbitMQ message (it could be interesting to map the incoming message subject to RabbitMQ message binding key)
Also, the AMQP 1.0 incoming message contains Application Properties which is a free map of key/value pairs. It should be also interesting to translate the incoming AMQP 1.0 message Application Properties in AMQP 0.9.1 messages headers in order to define routing in RabbitMQ Server

### Describe the solution you'd like

Perhaps in RabbitMQ configuration file
a sections concerning AMQP 1.0 incoming messages
ex : 
# Activate translation of AMQP 1.0 incoming message subject to amqp 0.9.1 message routing key
amqp1_0.map_subject_to_routing_key = true
# Allow to add AMQP 1.0 incoming message subject to amqp 0.9.1 message headers
amqp1_0.copy_subject_to_amqp091_headers = true
# Allow to copy AMQP 1.0 incoming message Application Properties (free map of key/values pairs) 
# to amqp 0.9.1 message headers
amqp1_0.copy_app_props_to_amqp091_headers = true

### Describe alternatives you've considered

My today alternative is to emplement an application which act as proxy in order to convert AMQP 1.0 incoming messages to AMQP 0.9.1 for RabbitMQ

### Additional context

Need to exchange between 2 systems. The first is a legacy system in a different organisation which is only able to exchange with IBM MQ.
And the target system is a new one, and the organisation don't have in minf to use IBM MQ. They are working at the integration level with RabbitMQ Clusters

**Comment by michaelklishin** [comment]
Translation of AMQP 1.0 Application Properties does not match very well to AMQP 0-9-1. There was a couple of attempts that varied in scope. #6273 is one example. The plan with #5077 is to move to AMQP 1.0 format internally.

Subject to routing key is a drastically simpler problem definition. @kjnilsson is there a specific reason why it is not currently mapped?

**Reply by DarkIkki** [reply]
For incoming AMQP 1.0 message Application Properties, perhaps allowing to list the Application Properties to copy, and in option the mapping rules.
For example :
=> in a simple way
amqp1_0.copy_app_props_to_amqp091_headers = content_type, content_encoding, entity_type
=> with mapping 
amqp1_0.copy_app_props_to_amqp091_headers = content_type:message_type, content_encoding:message_encoding, entity_type:data_type

could be also intersting for incoming AMQP 1.0 message Properties, like the "to" property
=> in a simple way
amqp1_0.copy_props_to_amqp091_headers = to 
=> with mapping 
amqp1_0.copy_props_to_amqp091_headers = to:amqp10_to

**Reply by DarkIkki** [reply]
I forgot, the will be intersting to inject this additional configuration at 2 differents levels :
1. In the configuration file for RabbitMQ AMQP 1.0 clients
2. In the Shovel definition when defining "src" or "dest" as with AMQP 1.0 protocol

**Reply by michaelklishin** [reply]
Do these examples not look awful to you? How can the same level of complexity fit Shovels where the only bit of configuration is a URI?

**Reply by michaelklishin** [reply]
I am first and foremost interested in looking into propagating subject to routing key. As for all other properties, dumping them as a map into a header is the best idea I have after looking at three or so alternatives.

So let's focus on the subject part.

**Reply by DarkIkki** [reply]
Hi,

When you said "As for all other properties, dumping them as a map into a header is the best idea I have" - you mean one prperty to one header entry ? Otherwise, it will not be possible to use header excha.ge for routine message.

Thanks

**Reply by michaelklishin** [reply]
It can be one property in one header of one property or one map header with all Application Properties grouped togther. Most users do not use header exchanges.

**Reply by DarkIkki** [reply]
Ok, we sometimes use the header exchanges.
It will be intersting to do the same in the other way :

AMQP 0.9.1 -> AMQP 1.0 :
Routine key -> Subject
Headers -> Application Properties

And : 
- the host part of the shovel target URI -> "to" Property of the AMQP 1.0 outgoing message
- the user part of the shovel target URI -> "user_id" Property of the AMQP 1.0 outgoing message 

 * In AMQP 1.0,  Propeties is a different segment than Application Properties. Some Server AMQP 1.0 implementation expect a value fo the "to" property.

**Reply by michaelklishin** [reply]
Like I said, I am only interested in solving the Routing key <=> Subject, and maybe the `to` property, translation for Shovels.

Everything else should wait until after https://github.com/rabbitmq/rabbitmq-server/pull/5077 is adopted (possibly in 3.13, possibly later).

**Comment by DarkIkki** [comment]
Ok, thank you.And do you have an idea of the version adoption for the subject and to mapping ?Thanks in advance
-------- Message d'origine --------De : Michael Klishin ***@***.***> Date : 08/04/2023  14:31  (GMT+01:00) À : rabbitmq/rabbitmq-server ***@***.***> Cc : DarkIkki ***@***.***>, Author ***@***.***> Objet : Re: [rabbitmq/rabbitmq-server] AMQP 1.0 [Shovel] Source (Discussion #7868) 
Like I said, I am only interested in solving the Routing key <=> Subject, and maybe the to property, translation for Shovels.
Everything else should wait until after #5077 is adopted (possibly in 3.13, possibly later).
